tower-http/src/validate_request.rs

```diff
@@ -0,0 +1,499 @@
+//!     .await?
+//!     .call(request)
+//!     .await?;
+//!
+//! assert_eq!(StatusCode::OK, response.status());
+//!
+//! // Requests with an invalid value get a `406 Not Acceptable` response
+//! let request = Request::builder()
+//!     .header(ACCEPT, "text/strings")
+//!     .body(Body::empty())
+//!     .unwrap();
+//!
+//! let response = service
+//!     .ready()
+//!     .await?
+//!     .call(request)
+//!     .await?;
+//!
+//! assert_eq!(StatusCode::NOT_ACCEPTABLE, response.status());
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! Custom validation can be made by implementing [`ValidateRequest`]:
+//!
+//! ```
+//! use tower_http::validate_request::{ValidateRequestHeaderLayer, ValidateRequest};
+//! use hyper::{Request, Response, Body, Error};
+//! use http::{StatusCode, header::ACCEPT};
+//! use tower::{Service, ServiceExt, ServiceBuilder, service_fn};
+//!
+//! #[derive(Clone, Copy)]
+//! pub struct MyHeader { }
+//!
+//! impl<B> ValidateRequest<B> for MyHeader {
+//!     type ResponseBody = Body;
+//!
+//!     fn validate(
+//!         &mut self,
+//!         request: &mut Request<B>,
+//!     ) -> Result<(), Response<Self::ResponseBody>> {
+//!         # unimplemented!()
+//!     }
+//! }
+//!
+//! async fn handle(request: Request<Body>) -> Result<Response<Body>, Error> {
+//!     Ok(Response::new(Body::empty()))
+//! }
+//!
+//! 
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let service = ServiceBuilder::new()
+//!     // Validate requests using `MyHeader`
+//!     .layer(ValidateRequestHeaderLayer::custom(MyHeader{}))
+//!     .service_fn(handle);
+//! # Ok(())
+//! # }
+//! ```
+//!
+//! Or using a closure:
+//!
+//! ```
+//! use tower_http::validate_request::{ValidateRequestHeaderLayer, ValidateRequest};
+//! use hyper::{Request, Response, Body, Error};
+//! use http::{StatusCode, header::ACCEPT};
+//! use tower::{Service, ServiceExt, ServiceBuilder, service_fn};
+//!
+//! async fn handle(request: Request<Body>) -> Result<Response<Body>, Error> {
+//!     # todo!();
+//!     // ...
+//! }
+//!
+//! # #[tokio::main]
+//! # async fn main() -> Result<(), Box<dyn std::error::Error>> {
+//! let service = ServiceBuilder::new()
+//!     .layer(ValidateRequestHeaderLayer::custom(|request: &mut Request<Body>| {
+//!         // Validate the request
+//!         # Ok::<_, Response<Body>>(())
+//!     }))
+//!     .service_fn(handle);
+//! # Ok(())
+//! # }
+//! ```
+
+use http::{
+    header::{self, HeaderValue},
+    Request, Response, StatusCode,
+};
+use http_body::Body;
+use pin_project_lite::pin_project;
+use std::{
+    fmt,
+    future::Future,
+    marker::PhantomData,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower_layer::Layer;
+use tower_service::Service;
+
+/// Layer that applies [`ValidateRequestHeader`] which validates all requests using the
+/// [`ValidateRequest`] header.
+#[derive(Debug, Clone)]
+pub struct ValidateRequestHeaderLayer<T> {
+    valid: T,
```

Comment:
Calling this `valid` makes it sound like it should be a `bool`. Maybe `validate` is more clear?

```suggestion
    validate: T,
```